mainmatter/introduction/pipeline/index.tex

```diff
@@ -0,0 +1,96 @@
+    \path[arrow] (gs) -- node{Primitives} (r);
+    \path[arrow] (r)  -- node{Fragment} (ps);
+    \path[arrow] (ps) -- node{Color} (om);
+    
+    \path[arrow] (m.west|-ia) -- (ia);
+    \path[arrow] (m.west|-vs) -- (vs);
+    \path[arrow] (m.west|-pa) -- (pa);
+    \path[arrow] (m.west|-t) -- (t);
+    \path[arrow] (m.west|-gs) -- (gs);
+    \path[arrow] (m.west|-ps) -- (ps);
+    \path[arrow, <->, shorten <=1pt] (m.west|-om) -- (om);
+  \end{tikzpicture}
+  \caption{A view of the graphics pipeline. The orange boxes are fixed functions, the blue boxes with rounded corners are shaders that the developer can replace with their own program, and the large green box is shared memory. Dashed shader stages are optional.}
+  \label{fig:pipeline01}
+\end{figure}
+
+The following paragraph goes through the different stages in the graphics pipeline, show in Figure \ref{fig:pipeline01}.
+
+\paragraph{The Input Assembler} is the first stage. 
+It reads the user provided data and assembles the data into structures that it sends to the next stage in the pipeline.
+These structures include primitive types in the shader languages, such as floats, vectors, matrices, etc.
+
+\paragraph{The Vertex Shader} is the first interchangeable stage in the pipeline.
+Its main responsibility is to take a set of output from the Input Assembler and translate it into vertices in clip-space.
+Clip-space is a space where all verticies that should be displayed are within the x and y coordinates -1 to 1 and z coordinate 0 to 1 or -1 to 1 depending on the \gls{API}.
+It also calculate values that later stages in the pipeline uses.
+
+\paragraph{The Primitive Constructor} takes the vertices and constructs primitives.
+A primitive is either a point, a line, or a triangle.
+Other primitives exists but are \gls{API} specific.
+To enable reuse of vertices, a index buffer can be provided to tell the assembler which vertices makes a primitive.
+The primitive constructor also performs the necessary clipping so that off-screen vertices are not rendered.
+
+\paragraph{The Tessellation Stage} is the first optional stage.
+Its responsibility is to tessellate the primitive, which means dividing the primitive into multiple smaller primitives.
+In the figure it is simplified.
+The tessellation stage consists of three elements: two shader stages, and a fixed function between them.
+The first shader describes how much tessellation should be done, and how.
+The fixed function the generates the desired result and the last shader modifies the generated primitives.
+This modification could include moving vertices according to a mathematical formula, or a texture, to show details that was otherwise not present in the model.
+
+\paragraph{The Geometry Shader} takes in a single primitive and outputs any number of primitives.
+This is used to generate particles from points, or hair and grass on surfaces.
+It can also be used to mask some primitives if necessary.
```

Comment:
Can an example be given of whhen you would want to mask primitives?

Reply:
No

Reply:
No